Joenz Kim-Chan iP

joenzkc opened the pull request:
# DukeNoob
> "They said your mind should be for having ideas, but I really do not have ideas" - Myself ([source](https://www.youtube.com/watch?v=dQw4w9WgXcQ))


DukeNoob is just... Duke
- text-based
- easy to learn (I guess)
- ~~FAST~~ NOT THAT FAST

All you need to do is: 
1. download it from [ here ]( https://github.com/joenzkimchan/ip/releases/tag/A-Jar )
2. open command line, cd to the directory where the file is stored in, and type in the command "java -jar main.jar" to get started
3. add your tasks
4. let it manage your texts for you 😉
And it is **FREE**! (please pay me if you'd like tho)

Features:

- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Reminder (coming)
---
If you're a java programmer, you can use it to practice java too! 
``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

joenzkc committed:
Implement a skeletal version of Duke that starts by greeting the user, simply echoes the commands entered by the user, and exits when the user types `bye`

joenzkc committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

joenzkc committed:
Add the ability to mark tasks as done. Change from String to Task class.

joenzkc committed:
Add support for tracking three types of tasks:

    ToDos: tasks without any date/time attached to it e.g., visit new theme park
    Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
    Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm

joenzkc committed:
Test using the I/O redirection technique

joenzkc committed:
Error handling was already implemented during the earlier stages.
Add support for deleting tasks from the list.

joenzkc committed:
Add 2 more test cases and update I/O redirection

joenzkc committed:
Add save and load functionality.

joenzkc committed:
Update Event and Deadline to take in DateTime rather than strings

joenzkc committed:
Merge branch 'branch-Level-7'

joenzkc committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

joenzkc committed:
Refactor code to include more OOP
--------------------------------------------------
- Add Command classes
- Abstract print statements to Ui
- Add Parser class
- Refactor Duke class

joenzkc committed:
Divide classes into packages
-----------------------------------------
1. duke.command
2. duke.task

joenzkc committed:
Add JUnit tests to test the behaviour of parser

joenzkc committed:
Add manifest file for jar creation

joenzkc committed:
Update all javadocs to all classes.

joenzkc committed:
Tweak code to comply with coding standard.

joenzkc committed:
Add javadocs for tasks.

joenzkc committed:
Give users a way to find tasks by searching for a keyword.

joenzkc committed:
Merge branch 'branch-A-JavaDoc'

joenzkc committed:
Merge branch 'branch-A-CodingStandard'

joenzkc committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskList.java

joenzkc committed:
Merge remote-tracking branch 'origin/add-gradle-support'

joenzkc committed:
Add checkstyles and update build.gradle

lethiciars commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,59 @@
+
+/**
+ * This class represents Duke who runs the main program.
+ */
```
Perhaps a short description of what Duke/the main program does could be included?

lethiciars commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,112 @@
+package duke;
+
+import duke.command.*;
```
Perhaps the imported classes could be listed explicitly?

lethiciars commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.Objects;
+import java.util.Scanner;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
```
Perhaps the ordering of the import statements could be changed to group the java.time.* packages together?

lethiciars commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,122 @@
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.time.format.DateTimeParseException;
```
Perhaps the ordering of the import statements could be changed to group the java.time.* packages together?

lethiciars commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * This class represents the Command when the user types "event" validly.
+ */
+public class EventCommand extends Command{
```
Should there be a whitespace between Command and the bracket?

lethiciars reviewed (commented):
Neat, Overall adheres to most Java coding standards :-)

joenzkc committed:
Add GUi for Duke.

joenzkc committed:
Merge branch 'branch-Level-10'

itzblim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,49 @@
+import duke.Parser;
```
It could be helpful to include these files within the Duke package?

itzblim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,49 @@
+
+    /**
+     * Generates a response based off a user input.
+     *
```
May be helpful to include the parameter input here?

itzblim commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,121 @@
+        Command output;
+        output = new EmptyCommand();
+        switch (command) {
+        case "welcome":
```
I appreciate the use of the switch statement here!

itzblim commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,121 @@
+            output = new UnexpectedCommand("");
+            break;
+        }
+        return output;
```
This method could perhaps be shortened if every command was handled separately by individual methods? Up to you ultimately!

itzblim commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,249 @@
+     * @param taskList the current task list.
+     * @return the tasks in the list.
+     */
+    public static String listTasks(ArrayList<Task> taskList) {
```
Perhaps this method could be part of the Tasklist class instead?

itzblim commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,249 @@
+     * @param task task to be deleted.
+     * @return message when task is deleted.
+     */
+    public static String deleteTask(Task task) {
```
It might be helpful to clarify the method name to reflect that this method returns the delete task string rather than actually deleting the task?

itzblim commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,249 @@
+     *
+     * @return array out of bounds message.
+     */
+    public static String arrayIndexOutOfBoundsExceptionMessage() {
```
Perhaps another possible way to achieve the same functionalities as these multiple methods might be to declare these strings as class constants and then refer to them? (That might reduce the number of methods listed here!)

itzblim commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,35 @@
+package duke.command;
```
I think it's really good that you broke the different commands into their own classes!

itzblim reviewed (commented):
Overall i think you did really well to ensure the code quality of the project! Just a few minor changes you might want to consider! :)

joenzkc committed:
Add assertions to Duke.

Duke currently has some implicit assumptions.

Let's,
* ensure that the assumptions are correct
* Parser command should not be null, and the storage filepath should be valid

joenzkc committed:
Fix visual bug where Duke message will cut off if too long

Currently when the list gets too long, Duke will be only able to show 4 of the tasks

Let's,
* fix the dialog box to fit more tasks

joenzkc committed:
Improve code quality

There is quite a lot of dead code present in the code.

Let's delete the dead code.

joenzkc committed:
Merge pull request #1 from joenzkimchan/A-CodeQuality

Improve code quality

joenzkc committed:
Merge pull request #2 from joenzkimchan/A-Assertions

Add assertions to Duke

joenzkc committed:
Add support for tagging events and refactor some code

Some people would like to tag events

Let's add this functionality

joenzkc committed:
Fix printing format for tags

Previously tags were inconsistent, for ToDos it would print tags last, and for events and deadlines it would print the date at the back.

Let's make it consistent by printing tags at the back

joenzkc committed:
Merge pull request #3 from joenzkimchan/branch-C-Tagging

Add tagging functionality

joenzkc committed:
Fix delete task bug

joenzkc committed:
Merge branch 'master' of https://github.com/joenzkimchan/ip

joenzkc committed:
Set scene name to Duke and add screenshot

joenzkc committed:
Add README.md

joenzkc committed:
Set theme jekyll-theme-minimal

joenzkc committed:
Set theme jekyll-theme-cayman

joenzkc committed:
Add jemoji plugin to yml

joenzkc committed:
Set theme jekyll-theme-minimal

joenzkc committed:
Fix javadoc headers not starting with verbs

joenzkc committed:
Reduce lines of code in Parser and Storage methods

Currently the abstraction level in Parser and Storage could be higher.
Lets abstract out some of the code to make the method more understandable.

joenzkc committed:
Merge branch 'branch-A-CodeQuality'

joenzkc committed:
Merge branch 'master' of https://github.com/joenzkimchan/ip

joenzkc committed:
Edit README with link of release

joenzkc committed:
Update README code blocks

joenzkc committed:
Add blank space in README for better readibility

joenzkc committed:
Add missing javadocs
